invokeai/backend/model_manager/download/base.py

```diff
@@ -0,0 +1,281 @@
+# Copyright (c) 2023 Lincoln D. Stein and the InvokeAI Development Team
+"""Abstract base class for a multithreaded model download queue."""
+
+from abc import ABC, abstractmethod
+from enum import Enum
+from functools import total_ordering
+from pathlib import Path
+from typing import Callable, List, Optional, Union
+
+import requests
+from pydantic import BaseModel, Field
+from pydantic.networks import AnyHttpUrl
+
+from invokeai.app.services.config import InvokeAIAppConfig
+
+# Used to distinguish between repo_id sources and URL sources
+REPO_ID_RE = r"^[\w-]+/[.\w-]+$"
+HTTP_RE = r"^https?://"
+
+
+class DownloadJobStatus(str, Enum):
+    """State of a download job."""
+
+    IDLE = "idle"  # not enqueued, will not run
+    ENQUEUED = "enqueued"  # enqueued but not yet active
+    RUNNING = "running"  # actively downloading
+    PAUSED = "paused"  # previously started, now paused
+    COMPLETED = "completed"  # finished running
+    ERROR = "error"  # terminated with an error message
+    CANCELLED = "cancelled"  # terminated by caller
+
+
+class UnknownJobIDException(Exception):
+    """Raised when an invalid Job is referenced."""
+
+
+class ModelSourceMetadata(BaseModel):
+    """Information collected on a downloadable model from its source site."""
+
+    name: Optional[str] = Field(description="Human-readable name of this model")
+    author: Optional[str] = Field(description="Author/creator of the model")
+    description: Optional[str] = Field(description="Description of the model")
+    license: Optional[str] = Field(description="Model license terms")
+    thumbnail_url: Optional[AnyHttpUrl] = Field(description="URL of a thumbnail image for the model")
+    tags: Optional[List[str]] = Field(description="List of descriptive tags")
+
+
+DownloadEventHandler = Callable[["DownloadJobBase"], None]
+
+
+@total_ordering
+class DownloadJobBase(BaseModel):
+    """Class to monitor and control a model download request."""
```

Comment:
All model install jobs inherit from this model, but this model appears to a number of properties that are not relevant for all types of installs. For example, `access_token` and `subqueue` are only for `RepoID`,  `bytes` is irrelevant for `Path` installs, etc.

The result is that the narrowed classes `ModelInstallPathJob` contain a lot of extra stuff.

I think this should be restructured so each type of job has only the stuff it needs.

Reply:
I was already planning to do this. Will fix.